File: CPP/Creational/Builder/Example/ConcreteBuilder.cpp
//
// Created by ray on 4/16/25.
//
#include "ConcreteBuilder.h"

ConcreteBuilder::ConcreteBuilder() {
    product = new Product();
}

ConcreteBuilder::~ConcreteBuilder() {
    delete product;
}

void ConcreteBuilder::buildPartA() {
    product->addPart("PartA");
}

void ConcreteBuilder::buildPartB() {
    product->addPart("PartB");
}

Product* ConcreteBuilder::getResult() {
    return product;
}

File: PHP/Creational/Builder/Example/ConcreteBuilder.php
<?php

include_once 'Product.php';
include_once 'Builder.php';
class ConcreteBuilder implements Builder
{

    private Product $product;

    public function __construct()
    {
        $this->product = new Product();
    }
    public function buildPartA(): void
    {
        $this->product->addPart("PartA");
    }

    public function buildPartB(): void
    {
        $this->product->addPart("PartB");
    }

    public function getResult(): Product
    {
        return $this->product;
    }
}